src/dynamics/mpas/dyn_comp.F90

```diff
@@ -139,13 +175,575 @@ pure function sec_to_hour_min_sec(sec) result(hour_min_sec)
         end function sec_to_hour_min_sec
     end subroutine dyn_readnl
 
+    !> Initialize MPAS dynamical core by one of the following:
+    !> 1. Setting analytic initial condition;
+    !> 2. Reading initial condition from a file;
+    !> 3. Restarting from a file.
+    !> (KCW, 2024-05-28)
+    !
     ! Called by `cam_init` in `src/control/cam_comp.F90`.
     subroutine dyn_init(cam_runtime_opts, dyn_in, dyn_out)
-        type(runtime_options), intent(in)  :: cam_runtime_opts
-        type(dyn_import_t),    intent(out) :: dyn_in
-        type(dyn_export_t),    intent(out) :: dyn_out
+        type(runtime_options), intent(in) :: cam_runtime_opts
+        type(dyn_import_t), intent(in) :: dyn_in
+        type(dyn_export_t), intent(in) :: dyn_out
+
+        character(*), parameter :: subname = 'dyn_comp::dyn_init'
+        character(std_name_len), allocatable :: constituent_name(:)
+        integer :: coupling_time_interval
+        integer :: i
+        integer :: ierr
+        logical, allocatable :: is_water_species(:)
+        type(file_desc_t), pointer :: pio_init_file => null()
+        type(file_desc_t), pointer :: pio_topo_file => null()
+
+        allocate(constituent_name(num_advected), stat=ierr)
+        call check_allocate(ierr, 'dyn_init', 'constituent_name', 'dyn_comp', __LINE__)
+
+        allocate(is_water_species(num_advected), stat=ierr)
+        call check_allocate(ierr, 'dyn_init', 'is_water_species', 'dyn_comp', __LINE__)
+
+        do i = 1, num_advected
+            constituent_name(i) = const_name(i)
+            is_water_species(i) = const_is_water_species(i)
+        end do
```

Comment:
The two variables are used to call `dyn_mpas_define_scalar` [[1]](https://github.com/ESCOMP/CAM-SIMA/blob/724d5c859556d4ed9bc13de039b574f0810b7252/src/dynamics/mpas/driver/dyn_mpas_subdriver.F90#L855), which informs MPAS, at run-time, about **all** constituent names and their corresponding waterness.

Indeed, I can choose to pass `const_name` and `const_is_water_species` as procedure pointers and call them from MPAS side. But, this will involve assumptions about their procedure interfaces, which is not ideal.

I pushed https://github.com/ESCOMP/CAM-SIMA/pull/267/commits/724d5c859556d4ed9bc13de039b574f0810b7252 to directly use them elsewhere.